### PR TITLE
Add delivery details to orders api

### PR DESCRIPTION
**Pull request by ethatcher83**
Add delivery details DTO
Add delivery details mapper
Add delivery details to basket controller
Add delivery details integration and unit tests

Resolves: PCI-715

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java [hunk 1]
```diff
@@ -86,6 +96,35 @@ public BasketController(final BasketMapper mapper,
         return ResponseEntity.status(HttpStatus.OK).body(addToBasketResponseDTO);
     }
 
+    @PatchMapping("${uk.gov.companieshouse.orders.api.basket}")
+    public ResponseEntity<?> addDeliveryDetailsToBasket(final @Valid @RequestBody AddDeliveryDetailsRequestDTO addDeliveryDetailsRequestDTO,
+                                                                 HttpServletRequest request,
+                                                                 final @RequestHeader(REQUEST_ID_HEADER_NAME) String requestId){
+        trace("ENTERING addDeliveryDetailToBasket(" + addDeliveryDetailsRequestDTO + ")", requestId);
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java]: Minor point: method name has changed.

File: src/main/java/uk/gov/companieshouse/orders/api/validator/DeliveryDetailsValidator.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    public DeliveryDetailsValidator() {
+    }
+
+    public List<String> getValidationErrors(final AddDeliveryDetailsRequestDTO addDeliveryDetailsRequestDTO) {
+        List<String> errors = new ArrayList<>();
+
+        if(addDeliveryDetailsRequestDTO.getPostalCode().isEmpty() && addDeliveryDetailsRequestDTO.getRegion().isEmpty()) {
+            errors.add("Post code or Region is required");
+        }
+        if(addDeliveryDetailsRequestDTO.getAddressLine1().isEmpty()) {
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/orders/api/validator/DeliveryDetailsValidator.java]: Could the validations from here downwards have been achieved with the annotations on the DTO?

File: src/test/java/uk/gov/companieshouse/orders/api/controller/BasketControllerIntegrationTest.java [hunk 3]
```diff
@@ -242,6 +260,62 @@ public void checkoutBasketReturnsBadRequestIfBodyIsPresent() throws Exception {
         assertEquals(0, checkoutRepository.count());
     }
 
+    @Test
+    @DisplayName("Add delivery details to the basket, if the basket exists")
+    public void addDeliveryDetailsToBasketIfTheBasketExists() throws Exception {
+        Basket basket = new Basket();
```
Comment (VaughanJackson123) [inline src/test/java/uk/gov/companieshouse/orders/api/controller/BasketControllerIntegrationTest.java]: Would there be value in a similar test for what happens when the basket does not exist?

**Comment by VaughanJackson123** [comment]
The ticket notes that the request will be tested using Postman. Have you considered adding Postman request(s) as part of this?

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java (hunk 1 above)
Comment (ethatcher83) [reply src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java]: Updated text

File: src/main/java/uk/gov/companieshouse/orders/api/validator/DeliveryDetailsValidator.java (hunk 2 above)
Comment (ethatcher83) [reply src/main/java/uk/gov/companieshouse/orders/api/validator/DeliveryDetailsValidator.java]: Removed validator tests

File: src/test/java/uk/gov/companieshouse/orders/api/controller/BasketControllerIntegrationTest.java (hunk 3 above)
Comment (ethatcher83) [reply src/test/java/uk/gov/companieshouse/orders/api/controller/BasketControllerIntegrationTest.java]: Added extra test

**Pull request merged**